perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/AppAutoRejectionScheduler.java

```diff
@@ -0,0 +1,252 @@
+	 * VO Manager), then rejects this application.
+	 *
+	 * @param vos eligible virtual organizations
+	 * @throws PerunException perun exception
+	 */
+	private void voApplicationsAutoRejection(List<Vo> vos) throws PerunException {
+		List<String> states = new ArrayList<>();
+		states.add("NEW");
+		states.add("VERIFIED");
+
+		for (Vo vo : vos) {
+			Attribute expiration = perun.getAttributesManagerBl().getAttribute(sess, vo, A_VO_APP_EXP_RULES);
+			if (expiration.getValue() != null) {
+				List<Application> applications = registrarManager.getApplicationsForVo(sess, vo, states);
+				rejectExpiredApplications(applications, expiration);
+			}
+		}
+	}
+
+	/**
+	 * Gets all existing groups and then checks all applications for this groups and if some application is expired
+	 * (according to expiration rules set by VO Manager), then rejects this application.
+	 *
+	 * @param groups eligible groups
+	 * @throws PerunException perun exception
+	 */
+	private void groupApplicationsAutoRejection(List<Group> groups) throws PerunException {
+		List<String> states = new ArrayList<>();
+		states.add("NEW");
+		states.add("VERIFIED");
+
+		for (Group group : groups) {
+			Attribute expiration = perun.getAttributesManagerBl().getAttribute(sess, group, A_GROUP_APP_EXP_RULES);
+			if (expiration.getValue() != null) {
+				List<Application> applications = registrarManager.getApplicationsForGroup(sess, group, states);
+				rejectExpiredApplications(applications, expiration);
+			}
+		}
+	}
+
+	/**
+	 * Compares date of last modification of application to values in expiration attribute and if finds expired application, then
+	 * rejects it.
+	 *
+	 * @param applications applications
+	 * @param expiration attribute with number of days to application expiration
+	 */
+	private void rejectExpiredApplications (List<Application> applications, Attribute expiration) {
+		Map<String, String> attrValue = expiration.valueAsMap();
+		for(Application application : applications) {
+			String date = application.getModifiedAt();
+			LocalDate modifiedAt = LocalDate.parse(date.substring(0, 10));
+			LocalDate now = getCurrentLocalDate();
+			if (application.getState() == Application.AppState.NEW) {
+				int expirationAppWaitingForEmail = Integer.parseInt(attrValue.get("emailVerification"));
+				if (now.minusDays(expirationAppWaitingForEmail).isAfter(modifiedAt)) {
+					String reasonForVo = "Your application to VO " + application.getVo().getName() + " was auto rejected, because you didn't verify your email address.";
+					String reasonForGroup = application.getGroup() == null ? "" : "Your application to group " + application.getGroup().getName() + " was auto rejected, because you didn't verify your email address.";
+					rejectWithReason(application, reasonForVo, reasonForGroup);
+					continue;
+				}
+			}
+			int expirationAppIgnoredByAdmin = Integer.parseInt(attrValue.get("ignoredByAdmin"));
```

Comment:
I would add the whole part with "ignoreByAdmin" option to the else branch (or `else if (Application.AppState.VERIFIED)`).

The reason is this (what could happen now):
It can happen that the "ignoredByAdmin" is set to less days than "emailVerification".
Then when the application is in state NEW but the "emailVerification" did not expired yet, it will auto-reject the application with the message saying that admin did not approve the application.
However, this is not true because the application was not even in the approval process yet (it was waiting for the email verification). Therefore, it couldn't (or shouldn't) be approved.

Reply:
But on the other hand, this is on Vo/Group admin, if he set it like this, isn't it? In the first task (https://homeproj.cesnet.cz/issues/6556) was written, that the second value of attribute ("ignoredByAdmin") shouldn't depend on state, so I think, that this possibility is desirable. Due to this I have the "continue" in the if statement, so if the number of days for both values is the same, then shouldnť happen the double try to reject the same application.

Reply:
I see. I think the best way to do is to ask the task requestor :) How should we solve it @vyskocilpavel ?

Reply:
It's possible to make a decision on the combination of both attributes?

1. use-case:
* emailVerification - 14 days
* ignoredByAdmin - unset
* then reject only application in state: "Waiting for mail verification" after 14 days

2. use-case:
* emailVerification - unset
* ignoredByAdmin - 30 days
* then reject all applications after 30 days

3. use-case:
* emailVerification - 14 days
* ignoredByAdmin - 30 days
* then reject all unverified applications after 14 days and all other applications after 30 days

Reply:
I have discussed it with Pavel and the result is following. If there are both limits set, we want to separately check NEW applications and separately VERIFIED applications. So the code should be constructed something like this:

```
if (application.getState() == Application.AppState.NEW && attrValue.containsKey("emailVerification")) {
    //process this situation
} else if (attrValue.containsKey("ignoredByAdmin")) {
    //process this situation
}
```

This should cover all possible use cases. Correct me if I am wrong. I could overlook something.